lib/logsource.c

```diff
@@ -44,10 +44,38 @@ log_source_wakeup(LogSource *self)
   msg_diagnostics("Source has been resumed", log_pipe_location_tag(&self->super));
 }
 
+static inline guint32
+_take_reclaimed_window(LogSource *self, guint32 window_size_increment)
+{
+  gssize old = atomic_gssize_sub(&self->window_size_to_be_reclaimed, window_size_increment);
+  gsize reclaimed_dec = window_size_increment;
+
+  if (old > 0)
+    {
+      if (old - window_size_increment >= 0)
+        {
+          window_size_increment = 0; // all the acks goes back to the common window pool
+        }
+      else
+        {
+          reclaimed_dec = window_size_increment - old;
```

Comment:
If I understand right, the `reclaimed_dec` variable stores the number of ACKs (window_size_increments), which we give to the dynamic pool.
We run to this `else` branch, when the `old` (window size needed for the dynamic pool) is less, than the ACKs we received. In this case we give back `old` to the dynamic pool, and not `window_size_increment - old`.
Please correct me, if I understand it wrong.

Reply:
I prepared a little refactor for this code also, with the problem fixed, too.
```
static inline guint32
_take_reclaimed_window(LogSource *self, guint32 window_size_increment)
{
  guint32 remaining_window_size_increment = window_size_increment;
  gssize old = atomic_gssize_sub(&self->window_size_to_be_reclaimed, window_size_increment);
  gboolean need_to_reclaim = (old > 0);

  if (!need_to_reclaim)
    return remaining_window_size_increment;

  remaining_window_size_increment = MAX(window_size_increment - old, 0);

  if (remaining_window_size_increment)
    atomic_gssize_add(&self->pending_reclaimed, old);
  else
    atomic_gssize_add(&self->pending_reclaimed, window_size_increment);

  return remaining_window_size_increment;
}
```

Reply:
We can further simplify this by replacing
```
  if (remaining_window_size_increment)
    atomic_gssize_add(&self->pending_reclaimed, old);
  else
    atomic_gssize_add(&self->pending_reclaimed, window_size_increment);
```
with
```
  atomic_gssize_add(&self->pending_reclaimed, window_size_increment - remaining_window_size_increment);
```

Reply:
@altilla:  you are right, but I think that else branch is not possible : cannot get more ACKs (I need to check again with pen&paper).

Reply:
@altilla:  you are right, but I think that else branch is not possible : cannot get more ACKs (I need to check again with pen&paper).